src/main/java/org/datadog/jenkins/plugins/datadog/traces/StepDataAction.java

```diff
@@ -11,19 +12,27 @@
 
 /**
  * Keeps the Step data during a certain Run.
+ *
+ * Note: We need to synchronize with the run instance because in parallel pipelines the WorkflowRun.save() method
+ * may raise a ConcurrentModificationException if the action is being persisted and it's modified during the process.
+ * We synchronize based on the run instance because the WorkflowRun.save() method synchronize on this.
  */
 public class StepDataAction extends InvisibleAction implements Serializable {
 
     private static final long serialVersionUID = 1L;
 
     private final Map<String, StepData> stepDataByDescriptor = new HashMap<>();
 
-    public StepData put(final FlowNode flowNode, final StepData stepData) {
-        return stepDataByDescriptor.put(flowNode.getId(), stepData);
+    public StepData synchronizedPut(final Run<?,?> run, final FlowNode flowNode, final StepData stepData) {
+        synchronized (run){
+            return stepDataByDescriptor.put(flowNode.getId(), stepData);
+        }
     }
 
-    public StepData get(final FlowNode flowNode) {
-        return stepDataByDescriptor.get(flowNode.getId());
+    public StepData synchronizedGet(final Run<?,?> run, final FlowNode flowNode) {
+        synchronized (run){
```

Comment:
why are you sych'ing over run and not over the hashmap of the data action?

Reply:
Because the lock object is the `run` instance in the [WorkflowRun.save()](https://github.com/jenkinsci/workflow-job-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java#L1204) method. 
If we synchronize on the hashmap directly will not resolve the issue, because that is not the used lock object.